mm2src/mm2_main/src/lp_healthcheck.rs

```diff
@@ -0,0 +1,505 @@
+            }
+
+            fn visit_str<E>(self, value: &str) -> Result<PeerAddress, E>
+            where
+                E: serde::de::Error,
+            {
+                if value.len() > 100 {
+                    return Err(serde::de::Error::invalid_length(
+                        value.len(),
+                        &"peer id cannot exceed 100 characters.",
+                    ));
+                }
+
+                Ok(mm2_libp2p::PeerId::from_str(value)
+                    .map_err(serde::de::Error::custom)?
+                    .into())
+            }
+
+            fn visit_string<E>(self, value: String) -> Result<PeerAddress, E>
+            where
+                E: serde::de::Error,
+            {
+                self.visit_str(&value)
+            }
+        }
+
+        deserializer.deserialize_str(PeerAddressVisitor)
+    }
+}
+
+impl HealthcheckMessage {
+    pub(crate) fn generate_message(
+        ctx: &MmArc,
+        target_peer: PeerAddress,
+        is_a_reply: bool,
+        expires_in_seconds: i64,
+    ) -> Result<Self, String> {
+        let p2p_ctx = P2PContext::fetch_from_mm_arc(ctx);
+        let sender_peer = p2p_ctx.peer_id().into();
+        let keypair = p2p_ctx.keypair();
+        let sender_public_key = keypair.public().encode_protobuf();
+
+        let data = HealthcheckData {
+            sender_peer,
+            sender_public_key,
+            target_peer,
+            expires_at: Utc::now().timestamp() + expires_in_seconds,
+            is_a_reply,
+        };
+
+        let signature = try_s!(keypair.sign(&try_s!(data.encode())));
+
+        Ok(Self { signature, data })
+    }
+
+    pub(crate) fn is_received_message_valid(
+        &self,
+        my_peer_address: PeerAddress,
+        healthcheck_config: &HealthcheckConfig,
+    ) -> bool {
+        let now = Utc::now().timestamp();
+        let remaining_expiration_seconds = u64::try_from(self.data.expires_at - now).unwrap_or(0);
+
+        if remaining_expiration_seconds == 0 {
+            log::debug!(
+                "Healthcheck message is expired. Current time in UTC: {now}, healthcheck `expires_at` in UTC: {}",
+                self.data.expires_at
+            );
+            return false;
+        } else if remaining_expiration_seconds > healthcheck_config.message_expiration_secs {
+            log::debug!(
+                "Healthcheck message have too high expiration time.\nMax allowed expiration seconds: {}\nReceived message expiration seconds: {}",
+                self.data.expires_at,
+                remaining_expiration_seconds,
+            );
+            return false;
+        }
+
+        if self.data.target_peer != my_peer_address {
+            log::debug!(
+                "`target_peer` doesn't match with our peer address. Our address: '{}', healthcheck `target_peer`: '{}'.",
+                my_peer_address,
+                self.data.target_peer
+            );
+            return false;
+        }
```

Comment:
This really doesn't make the message invalid, but only not directed to us.
This plays badly with...